keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+For more background and guidance on this topic, see the [interoperability](#tooling-interoperability) section.
+
+ApplySets MAY have an annotation extending their scope, which MUST be respected by all tools if present:
+
+```yaml
+applyset.k8s.io/additional-namespaces: <ns>[,] # OPTIONAL
+```
+
+The `applyset.k8s.io/additional-namespaces` annotation extends the scope of the ApplySet. By default, the scope of an ApplySet is its parent object's namespace. When the parent is cluster-scoped but refers to namespace kinds (see below), or when the set spans multiple namespaces (which is not recommended, but allowed), this annotation can be used to extend the ApplySet's scoped to the listed namespaces. As cross-namespace ApplySets are not particularly encouraged, we do not currently optimize this further.  In particular, we do not specify the GKs per namespace.  We can add more annotations in future should the need arise.
+
+The value of this annotation is a comma-separated list of the names of the namespaces (other than the ApplySet namespace) in which objects are found, for example `default,kube-system,ns1,ns2`. We reserve the empty value. As with `applyset.k8s.io/contains-group-kinds`, this list of namespaces must be sorted alphabetically, and should be minimal (other than transiently during ApplySet mutations).
+
+#### Optional "hint" annotations
+
+The ApplySet parent object MAY also have one or more of the following annotations that help tooling identify ApplySet members more efficiently. These are annotations instead of labels because annotations can be larger, and because we definitely do not need to select on these values.
+
+While the use of either of these annotations assists arbitrary tools in listing the ApplySet members, their use is not required by the specification, and tooling should not populate such annotations unless it believes itself to be the manager of an ApplySet.
+
+Most tools will likely want to have some GK/object identification mechanism along these lines for performance and permissions reasons, but “porcelain” tooling can continue to do using its existing mechanisms, be they more efficient or more powerful or just easier to continue to support.
+
+We may revisit this and converge on a single, mandatory hint annotation before this KEP enters beta. If feedback shows a need for it, we may also consider additional "hint" mechanisms, such as supporting a field selector on the CRD that identifies a strongly-typed list of member objects.
+
+When a tool that wants to list the members of an ApplySet and cannot determine the list of GKs (i.e. because neither hint annotation is used), it may support “discovery”, likely warning that a full cluster crawl is being attempted. Insufficient permissions errors are likely with such functionality, and when they are encountered, the tool should warn that the membership list may be incomplete.
+
+```yaml
+applyset.k8s.io/contains-group-kinds: <kind>.<group>[,]` # OPTIONAL
```

Comment:
Again, due to the inability to enforce the data format, I think using a field in duck-typed CRDs would be better.

Reply:
Consider storing group-resource instead of group-kind?

Reply:
re group-resource, yes, I think I agree that `<resource>.<group>` will likely read better than `<kind>.<group>`.  Is it _just_ a style choice though - do we know of any examples where it would matter?

Reply:
There are a lot of annotations with varying complex validation logic. Maybe that makes sense for alpha. 

But it would seem some of these issues raised can be addressed if we use a new metadata field or field + core type rather than a raft of annotations? This would avoid the adoption limitations for putting all this in a CRD. It would be empty for non-parent objects, so I understand that seems weird to put it everywhere.

Another option would be to reverse that, and have a core applyset type, and allow it to reference the tool-specific porcelain resource as an object ref. That also solves the discovery of applysets issue. Has that been considered?

Reply:
That's a fair point, a built-in type is an option also.  I think we're setting ourselves up to get there eventually if that's the right outcome.  We first want to get lots of user feedback on this approach, which means that we want users to use it for real on real clusters.  I don't think we get any feedback cycles with a built-in type or API changes, where once we ship alpha we are essentially locked-in.  But if this approach is generally useful for users, we can then maybe justify a dedicated ApplySet CRD (and I like the idea of that having a ref to a tool-specific object), or even a built-in type, but I think we need to prove it out with users first.

In other words I don't think we're ruling out something more typed or something more built-in with this proposal, and we're hopefully building a strong understanding of what the right solution is.